File: bad-version.php
<?php
/* The isBadVersion API is defined in the parent class VersionControl.
      public function isBadVersion($version){} */

class VersionControl
{
    function isBadVersion($n)
    {
        global $badversion;
        return ($n >= $badversion? true: false);
    }
}

class Solution extends VersionControl {
    /**
     * @param Integer $n
     * @return Integer
     */
    function firstBadVersion($n) {
        if( $this->isBadVersion($n) == false )
            return 0;
        else if( $this->isBadVersion(1) != false )
            return 1;

        $left = 0;
        $right = $n;
        $lastBadVersion = 0;

        $ret = 0;
        while( $left <= $right ) {
            $mid = $left + ceil( ($right - $left)/ 2 );
            echo "\nLeft: ".$left."; Right: ".$right."; Mid: ".$mid."; LastBedVersion: ".$lastBadVersiona;
            $bad = $this->isBadVersion($mid);
            if( $bad == false ) {
                $left = $mid + 1;
                if( $lastBadVersion> 0 && $lastBadVersion - $mid == 1 )
                    break;
            }
            else {
                $right = $mid - 1;
                $lastBadVersion = $mid;
            }
            echo "\n2: Left: ".$left."; Right: ".$right."; Mid: ".$mid."; LastBedVersion: ".$lastBadVersiona;
        }
        return $lastBadVersion;
    }
}

global $badversion, $force;
$badversion = 20000;

$solution = new Solution();
echo sprintf( "\n%d is the first bad version.", $solution->firstBadVersion(20001) );

File: bad-version.py
# The isBadVersion API is already defined for you.
# def isBadVersion(version: int) -> bool:

def isBadVersion(n):
    global badVersion
    if n>=badVersion:
        return True
    else:
        return False

class Solution:
    def firstBadVersion(self, n: int) -> int:
        if isBadVersion(n) == False:
            return False
        elif isBadVersion(1) != False:
            return 1

        left = 1
        right = n
        lastBadVersion = 0

        while( right >= left ):
            mid = left + ((right-left)// 2)

            if isBadVersion(mid) == False:
                left = mid + 1
                if lastBadVersion > 0 and lastBadVersion - mid == 1:
                    break
            else:
                right = mid - 1
                lastBadVersion = mid
        
        return lastBadVersion

badVersion = 2000

sol = Solution()
ret = sol.firstBadVersion(2001)
print(f"\n{ret} is the first bad version.")